test/cpp/microbenchmarks/bm_event_engine_run.cc

```diff
@@ -0,0 +1,220 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <atomic>
+#include <cmath>
+#include <memory>
+
+#include <benchmark/benchmark.h>
+
+#include <grpc/event_engine/event_engine.h>
+#include <grpcpp/impl/grpc_library.h>
+
+#include "src/core/lib/event_engine/common_closures.h"
+#include "src/core/lib/event_engine/default_event_engine.h"
+#include "src/core/lib/event_engine/promise.h"
+#include "test/core/util/test_config.h"
+#include "test/cpp/microbenchmarks/helpers.h"
+#include "test/cpp/util/test_config.h"
+
+namespace {
+
+using ::grpc_event_engine::experimental::AnyInvocableClosure;
+using ::grpc_event_engine::experimental::EventEngine;
+using ::grpc_event_engine::experimental::GetDefaultEventEngine;
+using ::grpc_event_engine::experimental::Promise;
+using ::grpc_event_engine::experimental::ResetDefaultEventEngine;
+
+void BM_EventEngine_RunLambda(benchmark::State& state) {
+  int cb_count = state.range(0);
+  std::atomic_int cnt{0};
+  Promise<bool> p{false};
+  auto cb = [&cnt, &p, cb_count]() {
+    if (++cnt == cb_count) p.Set(true);
+  };
+  auto engine = GetDefaultEventEngine();
+  for (auto _ : state) {
+    for (int i = 0; i < cb_count; i++) {
+      engine->Run(cb);
+    }
+    GPR_ASSERT(p.Get());
+    state.PauseTiming();
+    p.Reset();
+    cnt.store(0);
+    state.ResumeTiming();
+  }
+  ResetDefaultEventEngine();
+  state.SetItemsProcessed(cb_count * state.iterations());
+}
+BENCHMARK(BM_EventEngine_RunLambda)
+    ->Range(100, 4096)
+    ->MeasureProcessCPUTime()
+    ->UseRealTime();
+
+void BM_EventEngine_RunClosure(benchmark::State& state) {
+  int cb_count = state.range(0);
+  Promise<bool> p{false};
+  std::atomic_int cnt{0};
+  AnyInvocableClosure closure([&cnt, &p, cb_count]() {
+    if (++cnt == cb_count) {
+      p.Set(true);
+    }
+  });
+  auto engine = GetDefaultEventEngine();
+  for (auto _ : state) {
+    for (int i = 0; i < cb_count; i++) {
+      engine->Run(&closure);
+    }
+    GPR_ASSERT(p.Get());
+    state.PauseTiming();
+    p.Reset();
+    cnt.store(0);
+    state.ResumeTiming();
+  }
+  ResetDefaultEventEngine();
+  state.SetItemsProcessed(cb_count * state.iterations());
+}
+BENCHMARK(BM_EventEngine_RunClosure)
+    ->Range(100, 4096)
+    ->MeasureProcessCPUTime()
+    ->UseRealTime();
+
+void FanOutCallback(EventEngine* engine, std::atomic_int& cnt, int fanout,
+                    int depth, int limit, Promise<bool>& promise) {
+  if (++cnt == limit) {
```

Comment:
we could use atomic fetch_add with std::memory_order_acq_rel for faster atomic increments